File: javascript/032/main.js
/*
 * ProjectEuler 032
 * JavaScript / main.js
 * copyright (c) 2015 Susisu
 *
 * Node.js v5.0.0
 * `node main.js`
 */

"use strict";

// 1 * 1000 = 1000
// 10 * 100 = 1000

function main() {
    var list = [];
    var a, b, c;
    for (a = 1; a < 10; a++) {
        for (b = 1000; b < 10000; b++) {
            if (isPandigital(a, b, a * b)) {
                list.push(a * b);
            }
        }
    }
    for (a = 10; a < 100; a++) {
        for (b = 100; b < 1000; b++) {
            if (isPandigital(a, b, a * b)) {
                list.push(a * b);
            }
        }
    }
    list.sort((x, y) => x - y);
    var sum = 0;
    var pre = NaN;
    for (var i = 0; i < list.length; i++) {
        if (pre !== list[i]) {
            sum += list[i];
        }
        pre = list[i];
    }
    console.log(sum);
}

function isPandigital(a, b, c) {
    var str = a.toString() + b.toString() + c.toString();
    return str.split("").sort().join("") === "123456789"
}

main();

File: ocaml/032/main.ml
(*
    ProjectEuler 032
    OCaml / main.ml
    copyright (c) 2016 Susisu

    OCaml 4.02.3
    `ocamlfind ocamlopt -o bin/main -linkpkg -package core -thread main.ml`
*)

(*
    1 4 4
    2 3 4
*)

open Core

let is_pandigital a b p =
    let sa = string_of_int a in
    let sb = string_of_int b in
    let sp = string_of_int p in
    let s  = sa ^ sb ^ sp in
    let ss = Core_string.of_char_list (List.sort Char.compare (Core_string.to_list s)) in
    ss = "123456789"

let ps1 =
    let rec itr s a b =
        if a > 9 then s
        else if b > 9999 / a then itr s (a + 1) 1000
        else let p = a * b in
            if is_pandigital a b p then itr (p :: s) a (b + 1)
            else itr s a (b + 1)
    in itr [] 1 1000

let ps2 =
    let rec itr s a b =
        if a > 99 then s
        else if b > 9999 / a then itr s (a + 1) 100
        else let p = a * b in
            if is_pandigital a b p then itr (p :: s) a (b + 1)
            else itr s a (b + 1)
    in itr [] 10 100

let _ =
    let ps  = Core_list.dedup (ps1 @ ps2) in
    let sum = List.fold_left (+) 0 ps in
    print_int sum;
    print_newline ()
